library/core/src/main/java/com/google/android/exoplayer2/audio/Ac4Util.java

```diff
@@ -0,0 +1,269 @@
+/*
+ * Copyright (C) 2018 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.android.exoplayer2.audio;
+
+import com.google.android.exoplayer2.C;
+import com.google.android.exoplayer2.Format;
+import com.google.android.exoplayer2.drm.DrmInitData;
+import com.google.android.exoplayer2.util.MimeTypes;
+import com.google.android.exoplayer2.util.ParsableBitArray;
+import com.google.android.exoplayer2.util.ParsableByteArray;
+import java.nio.ByteBuffer;
+
+/**
+ * Utility methods for parsing AC-4 frames, which are access units in AC-4 bitstreams.
+ */
+public final class Ac4Util {
+
+  /**
+   * Holds sample format information as presented by a syncframe header.
+   */
+  public static final class SyncFrameInfo {
+
+    /**
+     * The sample mime type of the bitstream is {@link MimeTypes#AUDIO_AC4}.
+     */
+    public final String mimeType;
+    /**
+     * The bitstream version.
+     */
+    public final int bitstreamVersion;
+    /**
+     * The audio sampling rate in Hz.
+     */
+    public final int sampleRate;
+    /**
+     * The number of audio channels
+     */
+    public final int channelCount;
+    /**
+     * The size of the frame.
+     */
+    public final int frameSize;
+    /**
+     * Number of audio samples in the frame.
+     */
+    public final int sampleCount;
+
+    private SyncFrameInfo(
+        String mimeType,
+        int bitstreamVersion,
+        int channelCount,
+        int sampleRate,
+        int frameSize,
+        int sampleCount) {
+      this.mimeType = mimeType;
+      this.bitstreamVersion = bitstreamVersion;
+      this.channelCount = channelCount;
+      this.sampleRate = sampleRate;
+      this.frameSize = frameSize;
+      this.sampleCount = sampleCount;
+    }
+  }
+
+  /**
+   * The channel count of AC-4 stream.
+   */
+  // TODO: Parse AC-4 stream channel count.
+  public static final int CHANNEL_COUNT_2 = 2;
+  /**
+   * The header size for AC-4 parser. Only needs to be as big as we need to read, not the full
+   * header size.
+   */
+  public static final int HEADER_SIZE_FOR_PARSER = 16;
+  /**
+   * Number of audio samples in the frame. Defined in IEC61937-14:2017 table 5 and 6. This table
```

Comment:
Is the same information available in a freely-available ETSI specification? If so could you point me to it? I can see ETSI TS 103 190 2014-04 table 83 (subsection 4.3.3.2.6) has some numbers for frame_len_base that are similar but not exactly the same.

Reply:
Your quote is correct. I'd like to explain this table. For example, if frame_rate_index is 0, INTERNAL frame length is 1920. And decoder resampling ratio is 1001/1000*25/24. So EXTERNAL frame length is 1920*1001/1000*25/24 = 2002, which is same as the first line in my code.
The background of this table is AC-4 has feature to align the frame length between video and audio. For example, if video frame_rate is 23.976 fps, 2002 audio samples@48kHz will have same time duration as one video frame. (48000/2002 = 23.976)

Reply:
Got it. Thanks!